File: src/2025-02-23-deepclaude-try/javascript/exercises/practice/palindrome-products/palindrome-products.js
//
// This is only a SKELETON file for the 'Palindrome Products' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export class Palindromes {
  static generate({ minFactor, maxFactor }) {
    if (minFactor > maxFactor) {
      throw new Error('min must be <= max');
    }

    const palindromes = new Map();

    // Generate all products and find palindromes
    for (let i = minFactor; i <= maxFactor; i++) {
      for (let j = i; j <= maxFactor; j++) {
        const product = i * j;
        if (isPalindrome(product)) {
          if (!palindromes.has(product)) {
            palindromes.set(product, []);
          }
          palindromes.get(product).push([i, j]);
        }
      }
    }

    // Find smallest and largest palindromes
    let smallest = { value: null, factors: [] };
    let largest = { value: null, factors: [] };

    if (palindromes.size > 0) {
      const sortedProducts = Array.from(palindromes.keys()).sort((a, b) => a - b);
      smallest = {
        value: sortedProducts[0],
        factors: palindromes.get(sortedProducts[0])
      };
      largest = {
        value: sortedProducts[sortedProducts.length - 1],
        factors: palindromes.get(sortedProducts[sortedProducts.length - 1])
      };
    }

    return { smallest, largest };
  }
}

// Helper function to check if a number is palindrome
function isPalindrome(num) {
  const str = num.toString();
  return str === str.split('').reverse().join('');
}

File: src/2025-02-23-deepclaude-try/go/exercises/practice/palindrome-products/palindrome_products.go
package palindrome

import "errors"

type Product struct {
	Value          int
	Factorizations [][2]int
}

func Products(fmin, fmax int) (Product, Product, error) {
	if fmin > fmax {
		return Product{}, Product{}, errors.New("fmin > fmax")
	}

	products := make(map[int][][2]int)

	for a := fmin; a <= fmax; a++ {
		for b := a; b <= fmax; b++ {
			p := a * b
			if isPalindrome(p) {
				products[p] = append(products[p], [2]int{a, b})
			}
		}
	}

	if len(products) == 0 {
		return Product{}, Product{}, errors.New("no palindromes")
	}

	var minProduct, maxProduct int
	first := true
	for p := range products {
		if first {
			minProduct = p
			maxProduct = p
			first = false
		} else {
			if p < minProduct {
				minProduct = p
			}
			if p > maxProduct {
				maxProduct = p
			}
		}
	}

	minFactors := products[minProduct]
	maxFactors := products[maxProduct]

	return Product{minProduct, minFactors}, Product{maxProduct, maxFactors}, nil
}

func isPalindrome(n int) bool {
	if n < 0 {
		return false
	}
	reversed := 0
	original := n
	for n > 0 {
		reversed = reversed*10 + n%10
		n /= 10
	}
	return original == reversed
}
